File: src/chapters/numerical-calculus/content/code/simpson.r
# Composite Simpson's rule (n forced even) on [a, b].
simpson <- function(f, a, b, n = 100) {
  if (n %% 2 == 1) n <- n + 1
  x <- seq(a, b, length.out = n + 1)
  y <- f(x)
  h <- (b - a) / n
  odd <- seq(2, n, by = 2)      # interior odd-position points (1-indexed)
  even <- seq(3, n - 1, by = 2)
  h / 3 * (y[1] + y[n + 1] + 4 * sum(y[odd]) + 2 * sum(y[even]))
}

cat("int_0^1 e^x dx ~", simpson(exp, 0, 1, 100), " exact =", exp(1) - 1, "\n")

File: src/chapters/numerical-calculus/content/code/simpson.rs
// Composite Simpson's rule (n forced even).
fn simpson<F: Fn(f64) -> f64>(f: F, a: f64, b: f64, mut n: usize) -> f64 {
    if n % 2 == 1 { n += 1; }
    let h = (b - a) / n as f64;
    let mut s = f(a) + f(b);
    for i in 1..n { s += (if i % 2 == 1 { 4.0 } else { 2.0 }) * f(a + i as f64 * h); }
    h / 3.0 * s
}
fn main() {
    println!("int_0^1 e^x dx = {}", simpson(|x: f64| x.exp(), 0.0, 1.0, 100));
}
